pkg/k8s/diverts/types.go

```diff
@@ -0,0 +1,142 @@
+package diverts
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type IngressDivertSpec struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	Value     string `json:"value,omitempty"`
+}
+
+type ServiceDivertSpec struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	Port      int    `json:"port"`
+}
+
+type DeploymentDivertSpec struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type DivertSpec struct {
+	Ingress     IngressDivertSpec    `json:"ingress"`
+	FromService ServiceDivertSpec    `json:"fromService"`
+	ToService   ServiceDivertSpec    `json:"toService"`
+	Deployment  DeploymentDivertSpec `json:"deployment"`
+}
+
+type DivertStatus struct {
+	OriginalPort       int    `json:"originalPort"`
+	ProxyListenerPort  int    `json:"proxyListenerPort"`
+	HeaderID           string `json:"headerID"`
+	NodeID             string `json:"nodeID"`
+	DestinationAddress string `json:"destinationAddress"`
+	DestinationPort    int    `json:"destinationPort"`
+}
+
+type Divert struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   DivertSpec   `json:"spec"`
+	Status DivertStatus `json:"status"`
+}
+
+type DivertList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Divert `json:"items"`
+}
+
+// DeepCopyInto a deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *Divert) DeepCopyInto(out *Divert) {
+	*out = *in
+	out.TypeMeta = in.TypeMeta
+	in.ObjectMeta.DeepCopyInto(&out.ObjectMeta)
+	out.Spec = in.Spec
+	out.Status = in.Status
+}
+
+// DeepCopy a deepcopy function, copying the receiver, creating a new Divert.
+func (in *Divert) DeepCopy() *Divert {
+	if in == nil {
+		return nil
+	}
+	out := new(Divert)
+	in.DeepCopyInto(out)
+	return out
+}
+
+// DeepCopyObject a deepcopy function, copying the receiver, creating a new runtime.Object.
+func (in *Divert) DeepCopyObject() runtime.Object {
+	if c := in.DeepCopy(); c != nil {
+		return c
+	}
+	return nil
+}
+
+// DeepCopyInto a deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *DivertList) DeepCopyInto(out *DivertList) {
+	*out = *in
+	out.TypeMeta = in.TypeMeta
+	in.ListMeta.DeepCopyInto(&out.ListMeta)
+	if in.Items != nil {
+		in, out := &in.Items, &out.Items
+		*out = make([]Divert, len(*in))
+		for i := range *in {
+			(*in)[i].DeepCopyInto(&(*out)[i])
+		}
+	}
+}
+
+// DeepCopy a deepcopy function, copying the receiver, creating a new DivertList.
+func (in *DivertList) DeepCopy() *DivertList {
```

Comment:
can we add a unit test for this part? deepcopy might not do what we expect it to do in all the use cases we are handling.